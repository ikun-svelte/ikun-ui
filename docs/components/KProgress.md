---
title: KProgress
lang: en-US
---

# KProgress

Progress is used to show the progress of current operation.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/progress
```

```bash [yarn]
yarn add @ikun-ui/progress
```

```bash [npm]
npm install @ikun-ui/progress
```

:::

## Basic usage

<demo src="progress/basic.svelte" github="Progress"></demo>

## TextInside progress

In this case the percentage takes no additional space.

`strokeWidth` attribute decides the `width` of progress bar.

`textInside` attribute to put description inside the progress bar.

<demo src="progress/textInside.svelte" github="Progress"></demo>

## Color progress

`color` attribute decides the `color` of progress bar.

<demo src="progress/color.svelte" github="Progress"></demo>

## Circle progress

By setting `type="circle"`, you can get a circle style of progress easily.

<demo src="progress/circle.svelte" github="Progress"></demo>

## Dashboard progress

By setting `type="dashboard"`, you can get a dashboard style of progress easily.

<demo src="progress/dashboard.svelte" github="Progress"></demo>

## Progress Props

| Name        | Type                                         | Default    | Description                                                   |
| ----------- | -------------------------------------------- | ---------- | ------------------------------------------------------------- |
| percentage  | `number`                                     | `0`        | Percentage                                                    |
| status      | `primary / success / error / warning / info` | `primary`  | The current status of progress bar                            |
| type        | `line, circle, dashboard`                    | `line`     | The type of progress bar                                      |
| strokeWidth | `number`                                     | `6`        | The width of progress bar                                     |
| width       | `number`                                     | `126`      | The canvas width of circle progress                           |
| color       | `string`                                     | `-`        | The current color of progress bar                             |
| textInside  | `boolean`                                    | `false`    | Put description inside the progress bar (`type` is 'line')    |
| format      | `(v: number) => string`                      | `-`        | Custom text format                                            |
| duration    | `number`                                     | `0.28 (s)` | Control the animation duration of progress (`type` is 'line') |
| showText    | `boolean`                                    | `true`     | Whether to show percentage                                    |
| cls         | `string`                                     | `true`     | Additional class                                              |
| attrs       | `Record<string, string>`                     | `{}`       | Additional attributes                                         |

## Progress Slots

| Name    | Description        |
| ------- | ------------------ |
| default | Customized content |
