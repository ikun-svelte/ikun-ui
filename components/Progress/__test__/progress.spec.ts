import { tick } from 'svelte';
import { afterEach, beforeEach, describe, expect, test } from 'vitest';
import KProgress from '../src';
import KProgressSlots from './fixture/progress.slots.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KProgress', () => {
	test('props: line percentage', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const barElm = host.getElementsByClassName('k-progress--bar')[0] as HTMLDivElement;
		expect(barElm.style.width).toBe('10%');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: circle percentage', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				type: 'circle'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const circlElm = host.getElementsByTagName('circle')[1];
		const strokeDasharray = circlElm.getAttribute('stroke-dasharray')?.split(' ');
		expect(Number(strokeDasharray?.[0]) * 10).toBe(Number(strokeDasharray?.[1]));
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: dashboard percentage', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				type: 'dashboard'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const dashboardElm = host.getElementsByTagName('circle')[1];
		const strokeDasharrayBydashboard = dashboardElm.getAttribute('stroke-dasharray')?.split(' ');
		expect(((Number(strokeDasharrayBydashboard?.[0]) * 10) / 0.75).toFixed(2)).toEqual(
			Number(strokeDasharrayBydashboard?.[1]).toFixed(2)
		);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: line status', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				status: 'success',
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('k-progress--success')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: circle & dashboard status', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				status: 'success',
				type: 'circle'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('k-progress--stroke--success')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: line strokeWidth', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				strokeWidth: 520,
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const height = (host.children[0].children[0] as HTMLDivElement).style.height;
		expect(Number.parseFloat(height)).toBe(520);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: circle & dashboard strokeWidth', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				strokeWidth: 520,
				type: 'circle'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const circlElm = host.getElementsByTagName('circle')[0];
		const strokeWidth = circlElm.getAttribute('stroke-width');
		expect(Number(strokeWidth)).toBe(520);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: circle & dashboard width', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				width: 520,
				type: 'circle'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const Elm = host.children[0] as HTMLDivElement;
		expect(parseFloat(Elm.style.width)).toBe(520);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: color', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				color: '#000',
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const barElm = host.getElementsByClassName('k-progress--bar')[0] as HTMLDivElement;
		expect(barElm.style.backgroundColor).toEqual('rgb(0, 0, 0)');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: textInside & showText', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				textInside: true,
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const Elm = host.getElementsByClassName('k-progress--content__textInside')[0] as HTMLDivElement;
		const textContent = Elm.children[0].innerHTML;
		expect(parseFloat(textContent)).toBe(10);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: format', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				format: (v) => `我欲与君相知，长命无绝衰${v}`,
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes(`我欲与君相知，长命无绝衰10`)).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: duration', async () => {
		const instance = new KProgress({
			target: host,
			props: {
				percentage: 10,
				duration: 0.52,
				type: 'line'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		const barElm = host.getElementsByClassName('k-progress--bar')[0] as HTMLDivElement;
		expect(parseFloat(barElm.style.transitionDuration)).toEqual(0.52);
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: default', async () => {
		const instance = new KProgressSlots({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('defaultSlot')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
