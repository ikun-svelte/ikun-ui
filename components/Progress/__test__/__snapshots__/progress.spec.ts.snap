// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KProgress > props: circle & dashboard status 1`] = `"<div class=\\"k-progress__cd\\" style=\\"width: 126px; height: 126px;\\"><svg width=\\"126\\" height=\\"126\\" viewBox=\\"0 0 126 126\\"><circle cx=\\"63\\" cy=\\"63\\" r=\\"57\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"358.1415625092364 358.1415625092364\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 63, 63)\\" class=\\"k-progress--stroke--runway k-progress--stroke--runway__dark\\"></circle> <circle cx=\\"63\\" cy=\\"63\\" r=\\"57\\" stroke=\\"\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"0 358.1415625092364\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 63, 63)\\" class=\\"k-progress--bar__cd k-progress--stroke--success\\"></circle></svg> <div class=\\"k-progress--text__cd\\"><span>0%</span></div></div>"`;

exports[`Test: KProgress > props: circle & dashboard strokeWidth 1`] = `"<div class=\\"k-progress__cd\\" style=\\"width: 126px; height: 126px;\\"><svg width=\\"126\\" height=\\"126\\" viewBox=\\"0 0 126 126\\"><circle cx=\\"63\\" cy=\\"63\\" r=\\"-457\\" stroke-width=\\"520\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"-2871.415685381071 -2871.415685381071\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 63, 63)\\" class=\\"k-progress--stroke--runway k-progress--stroke--runway__dark\\"></circle> <circle cx=\\"63\\" cy=\\"63\\" r=\\"-457\\" stroke=\\"\\" stroke-width=\\"520\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"-287.1415685381071 -2871.415685381071\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 63, 63)\\" class=\\"k-progress--bar__cd k-progress--stroke--primary\\"></circle></svg> <div class=\\"k-progress--text__cd\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: circle & dashboard width 1`] = `"<div class=\\"k-progress__cd\\" style=\\"width: 520px; height: 520px;\\"><svg width=\\"520\\" height=\\"520\\" viewBox=\\"0 0 520 520\\"><circle cx=\\"260\\" cy=\\"260\\" r=\\"254\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"1595.929068023615 1595.929068023615\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 260, 260)\\" class=\\"k-progress--stroke--runway k-progress--stroke--runway__dark\\"></circle> <circle cx=\\"260\\" cy=\\"260\\" r=\\"254\\" stroke=\\"\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"159.5929068023615 1595.929068023615\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 260, 260)\\" class=\\"k-progress--bar__cd k-progress--stroke--primary\\"></circle></svg> <div class=\\"k-progress--text__cd\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: circle percentage 1`] = `"<div class=\\"k-progress__cd\\" style=\\"width: 126px; height: 126px;\\"><svg width=\\"126\\" height=\\"126\\" viewBox=\\"0 0 126 126\\"><circle cx=\\"63\\" cy=\\"63\\" r=\\"57\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"358.1415625092364 358.1415625092364\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 63, 63)\\" class=\\"k-progress--stroke--runway k-progress--stroke--runway__dark\\"></circle> <circle cx=\\"63\\" cy=\\"63\\" r=\\"57\\" stroke=\\"\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"35.81415625092364 358.1415625092364\\" stroke-dashoffset=\\"0\\" fill=\\"none\\" transform=\\"rotate(-90, 63, 63)\\" class=\\"k-progress--bar__cd k-progress--stroke--primary\\"></circle></svg> <div class=\\"k-progress--text__cd\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: color 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 10%; background-color: rgb(0, 0, 0); border-radius: 6px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: dashboard percentage 1`] = `"<div class=\\"k-progress__cd\\" style=\\"width: 126px; height: 126px;\\"><svg width=\\"126\\" height=\\"126\\" viewBox=\\"0 0 126 126\\"><circle cx=\\"63\\" cy=\\"63\\" r=\\"57\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"268.6061718819273 358.1415625092364\\" stroke-dashoffset=\\"-44.76769531365455\\" fill=\\"none\\" transform=\\"rotate(90, 63, 63)\\" class=\\"k-progress--stroke--runway k-progress--stroke--runway__dark\\"></circle> <circle cx=\\"63\\" cy=\\"63\\" r=\\"57\\" stroke=\\"\\" stroke-width=\\"6\\" stroke-linecap=\\"round\\" stroke-dasharray=\\"26.860617188192734 358.1415625092364\\" stroke-dashoffset=\\"-44.76769531365455\\" fill=\\"none\\" transform=\\"rotate(90, 63, 63)\\" class=\\"k-progress--bar__cd k-progress--stroke--primary\\"></circle></svg> <div class=\\"k-progress--text__cd\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: duration 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 10%; border-radius: 6px; transition-duration: 0.52s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: format 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 10%; border-radius: 6px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span>我欲与君相知，长命无绝衰10</span></div></div>"`;

exports[`Test: KProgress > props: line percentage 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 10%; border-radius: 6px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: line status 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--success\\" style=\\"width: 10%; border-radius: 6px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: line strokeWidth 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 520px; border-radius: 520px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 10%; border-radius: 520px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span>10%</span></div></div>"`;

exports[`Test: KProgress > props: textInside & showText 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 10%; border-radius: 6px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"><span>10%</span></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"></div></div>"`;

exports[`Test: KProgress > slots: default 1`] = `"<div class=\\"k-progress--container\\"><div class=\\"k-progress--base k-progress--base__dark\\" style=\\"width: 90%; height: 6px; border-radius: 6px;\\"><div class=\\"k-progress--runway\\"><div class=\\"k-progress--bar k-progress--primary\\" style=\\"width: 0%; border-radius: 6px; transition-duration: 0.28s; left: 0%;\\"><div class=\\"k-progress--content__textInside\\"></div></div></div></div> <div class=\\"k-progress--content__textOutside\\"><span class=\\"defaultSlot\\">🏀</span></div></div>"`;
